analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantInheritanceList.cs

```diff
@@ -145,50 +124,45 @@ private static void ReportRedundantInterfaces(SyntaxNodeAnalysisContext context,
-        private static MultiValueDictionary<INamedTypeSymbol, INamedTypeSymbol> GetImplementedInterfaceMappings(BaseListSyntax baseList, SemanticModel semanticModel) =>
+        private static ILookup<INamedTypeSymbol, INamedTypeSymbol> GetImplementedInterfaceMappings(BaseListSyntax baseList, SemanticModel semanticModel) =>
             baseList.Types
                     .Select(baseType => semanticModel.GetSymbolInfo(baseType.Type).Symbol as INamedTypeSymbol)
                     .WhereNotNull()
                     .Distinct()
-                    .Select(symbol => new Tuple<INamedTypeSymbol, ICollection<INamedTypeSymbol>>(symbol, symbol.AllInterfaces))
-                    .ToMultiValueDictionary(kv => kv.Item1, kv => kv.Item2);
+                    .ToLookup(x => x.AllInterfaces.AsEnumerable());
 
         private static bool TryGetCollidingDeclaration(INamedTypeSymbol declaredType,
                                                        INamedTypeSymbol interfaceType,
-                                                       MultiValueDictionary<INamedTypeSymbol, INamedTypeSymbol> interfaceMappings,
+                                                       ILookup<INamedTypeSymbol, INamedTypeSymbol> interfaceMappings,
                                                        out INamedTypeSymbol collidingDeclaration)
         {
-            var collisionMapping = interfaceMappings
-                .Where(i => i.Key.IsInterface())
-                .FirstOrDefault(v => v.Value.Contains(interfaceType));
-
-            if (collisionMapping.Key != null)
+            var collisionMapping = interfaceMappings.FirstOrDefault(x => x.Key.IsInterface() && x.Contains(interfaceType));
+            if (collisionMapping?.Key is not null)
             {
                 collidingDeclaration = collisionMapping.Key;
                 return true;
             }
 
-            var baseClassMapping = interfaceMappings.FirstOrDefault(i => i.Key.IsClass());
-            if (baseClassMapping.Key == null)
+            var baseClassMapping = interfaceMappings.FirstOrDefault(x => x.Key.IsClass());
+            if (baseClassMapping?.Key is null)
             {
                 collidingDeclaration = null;
                 return false;
             }
 
-            collidingDeclaration = baseClassMapping.Key;
-            return CanInterfaceBeRemovedBasedOnMembers(declaredType, interfaceType);
+            var canBeRemoved = CanInterfaceBeRemovedBasedOnMembers(declaredType, interfaceType);
+            collidingDeclaration = canBeRemoved ? baseClassMapping.Key : null;
+            return canBeRemoved;
         }
 
         private static bool CanInterfaceBeRemovedBasedOnMembers(INamedTypeSymbol declaredType, INamedTypeSymbol interfaceType)
         {
-            var allMembersOfInterface = interfaceType.AllInterfaces.Concat(new[] { interfaceType })
-                .SelectMany(i => i.GetMembers())
+            var allMembersOfInterface = AllInterfacesAndSelf(interfaceType)
```

Comment:
This can be single line. And I prefer `ToArray()` over `ToList()`, as we will not add/remove items to the result.

Reply:
I went for ImmutableArray and added a How we work entry.